### PR TITLE
cmake: enable windows build

### PR DESCRIPTION
**Pull request by matthiasdiener**
In principle ready to go.

Things that need careful review:
- [charmc](https://github.com/UIUC-PPL/charm/pull/2885/files#diff-048c96c9f30ba930d0a952dd7d22dbe2L1870) - for gen-cpm, changed a temp file from being a symlink to a copy, due to Win file system limitations.
- [unix2nt_cc](https://github.com/UIUC-PPL/charm/pull/2885/files#diff-6245319ce6778041aa6dffe988824323L341) - Several changes that make unix2nt_cc more generic (e.g. `-o path/file.o` instead of just `-o file.o` was not supported).

### CODE REVIEW

**Comment by evan-charmworks** [comment]
Tried this on MSYS2 locally:

```
$ ./build charm++ netlrts-win-x86_64
Build options for netlrts-win-x86_64:
  opt_CC = /g/Charmworks/charm/src/arch/win/unix2nt_cc
  opt_CXX = /g/Charmworks/charm/src/arch/win/unix2nt_cc
  opt_FC =
  opt_ampi_error_checking = 0
  opt_ampi_mpich_tests = 0
  opt_ampi_only = 0
  opt_build_shared = 0
  opt_ccs = 0
  opt_charmdebug = 0
  opt_charmpy = 0
  opt_compiler =
  opt_controlpoint = 0
  opt_cuda = 0
  opt_drone_mode = 0
  opt_error_checking = 0
  opt_extra_opts =
  opt_incdir =
  opt_lbtime_type = double
  opt_lbuserdata = 0
  opt_libdir =
  opt_lockless = 0
  opt_mempool_cutoff = 26
  opt_network = netlrts
  opt_numa = 0
  opt_omp = 0
  opt_parallel = -j1
  opt_prio_type = bitvec
  opt_production = Debug
  opt_pthreads = 0
  opt_pxshm = 0
  opt_qlogic = 0
  opt_randomized_msgq = 0
  opt_refnum_type = unsigned short
  opt_replay = 0
  opt_romio = 0
  opt_shrinkexpand = 0
  opt_smp = 0
  opt_stats = 0
  opt_suffix =
  opt_syncft = 0
  opt_task_queue = 0
  opt_tcp = 0
  opt_tracing = 0
  opt_tracing_commthread = 0
  opt_ucx_pmi =
  opt_zlib = 1
-- Building for: Visual Studio 16 2019
-- Selecting Windows SDK version 10.0.17763.0 to target Windows 10.0.18363.
-- The CXX compiler identification is MSVC 19.26.28806.0
-- The C compiler identification is MSVC 19.26.28806.0
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.26.28801/bin/Hostx64/x64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.26.28801/bin/Hostx64/x64/cl.exe - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.26.28801/bin/Hostx64/x64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.26.28801/bin/Hostx64/x64/cl.exe - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - not found
-- Found Threads: TRUE
-- Found OpenMP_C: -openmp (found version "2.0")
-- Found OpenMP_CXX: -openmp (found version "2.0")
-- Found OpenMP: TRUE (found version "2.0")
-- Charm++: No Fortran support.
CMake Error at CMakeLists.txt:67 (message):
  Unknown compiler MSVC


-- Configuring incomplete, errors occurred!
See also "G:/Charmworks/charm/netlrts-win-x86_64/CMakeFiles/CMakeOutput.log".
See also "G:/Charmworks/charm/netlrts-win-x86_64/CMakeFiles/CMakeError.log".

$ cmake --version
cmake version 3.17.3

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```

**Comment by matthiasdiener** [comment]
> -- Building for: Visual Studio 16 2019

Yeah, the problem is that cmake wants to configure for the VS build system, not Unix Makefiles with msys2, which is why I changed the CI back to cygwin.

**Comment by evan-charmworks** [comment]
Maybe under MSYS2, `./build` needs to pass `-G "MSYS Makefiles"` to CMake. https://cmake.org/cmake/help/latest/manual/cmake-generators.7.html

**Comment by matthiasdiener** [comment]
> Maybe under MSYS2, `./build` needs to pass `-G "MSYS Makefiles"` to CMake. https://cmake.org/cmake/help/latest/manual/cmake-generators.7.html

Yes, I tried that: https://github.com/UIUC-PPL/charm/pull/2885/commits/a940ca28a6ad15d28cc73228b3a3584cb4c93c1e 

There were other errors as far as I remember.

File: src/util/boost-context/CMakeLists.txt [hunk 1]
```diff
@@ -3,13 +3,13 @@ enable_language(C ASM)
...
 elseif(${CHARM_OS} STREQUAL "linux")
-    if(${CHARM_CPU} MATCHES "ppc")
+    if(${CHARM_CPU} MATCHES "ppc64le")
         set(ufcontext_sources jump_ppc64_sysv_elf_gas.S make_ppc64_sysv_elf_gas.S ontop_ppc64_sysv_elf_gas.S)
     else()
        set(ufcontext_sources jump_x86_64_sysv_elf_gas.S ontop_x86_64_sysv_elf_gas.S make_x86_64_sysv_elf_gas.S)
     endif()
 elseif(${CHARM_OS} STREQUAL "win")
-    if(MSVC)
+	if(${CMK_COMPILER} STREQUAL "msvc")
```
Comment (evan-charmworks) [inline src/util/boost-context/CMakeLists.txt:12]: Is the indentation correct here?

File: src/arch/win/unix2nt_cc [hunk 2]
```diff
@@ -390,7 +393,7 @@ fi
 #if [[ -n "$CL" ]]
 if [[ -n "$DOCOMPILE" ]]
 then
-	OPTS="$CL_OPTS $OUTPUT $CL"
+	OPTS="$CL_OPTS $OUTPUT $CL -I." # '-I.' in case inout file is a full path
```
Comment (evan-charmworks) [inline src/arch/win/unix2nt_cc:396]: `input`?

File: CMakeLists.txt [hunk 3]
```diff
@@ -550,6 +561,12 @@ if(UCX_PMI)
   configure_file(${CMAKE_SOURCE_DIR}/src/arch/${VDIR}/conv-mach-${UCX_PMI}.h include/ COPYONLY)
 endif()
 
+if(CHARM_OS STREQUAL "win")
+  configure_file(src/arch/win/unix2nt_cc                       bin/ COPYONLY)
+  configure_file(src/arch/win/unix2nt_ar                       bin/ COPYONLY)
+  configure_file(${CMAKE_SOURCE_DIR}/src/arch/win/unistd.h    include/ COPYONLY)
```
Comment (evan-charmworks) [inline CMakeLists.txt:567]: Spacing is inconsistent before `bin/` and `include/`

File: CMakeLists.txt [hunk 4]
```diff
@@ -250,8 +259,8 @@ set(CMK_USER_SUFFIX ".user")
 if(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
   set(CHARM_OS "darwin")
   set(CMK_MACOSX 1)
+elseif(CMAKE_SYSTEM_NAME STREQUAL "Windows" OR CMAKE_SYSTEM_NAME STREQUAL "CYGWIN")
   set(CMK_SHARED_SUF "dylib")
-elseif(CMAKE_SYSTEM_NAME STREQUAL "Windows")
```
Comment (evan-charmworks) [inline CMakeLists.txt:254]: the Windows line jumped above the `dylib` line here

File: src/util/boost-context/CMakeLists.txt (hunk 1 above)
Comment (matthiasdiener) [reply src/util/boost-context/CMakeLists.txt]: changed from tab to spaces.

File: src/arch/win/unix2nt_cc (hunk 2 above)
Comment (matthiasdiener) [reply src/arch/win/unix2nt_cc]: changed.

File: CMakeLists.txt (hunk 3 above)
Comment (matthiasdiener) [reply CMakeLists.txt]: Fixed.

File: CMakeLists.txt (hunk 4 above)
Comment (matthiasdiener) [reply CMakeLists.txt]: Fixed.

File: src/util/boost-context/CMakeLists.txt [hunk 5]
```diff
@@ -3,13 +3,13 @@ enable_language(C ASM)
 if (${CHARM_OS} STREQUAL "darwin")
     set(ufcontext_sources jump_x86_64_sysv_macho_gas.S ontop_x86_64_sysv_macho_gas.S make_x86_64_sysv_macho_gas.S)
 elseif(${CHARM_OS} STREQUAL "linux")
-    if(${CHARM_CPU} MATCHES "ppc")
+    if(${CHARM_CPU} MATCHES "ppc64le")
```
Comment (rbuch) [inline src/util/boost-context/CMakeLists.txt:6]: Are there any non-ppc64le targets that we should still be supporting?

File: CMakeLists.txt [hunk 6]
```diff
@@ -191,7 +195,12 @@ else()
   set(CMK_MULTICORE 0)
 endif()
 
-set(CMK_USE_ZLIB ${ZLIB})
+if(NOT ${CHARM_OS} STREQUAL "win")
```
[the remark this answers is not in the document]
Comment (evan-charmworks) [reply CMakeLists.txt:198]: It can be if building with GCC under MSYS2, but for our normal Visual Studio builds it might take special effort.

**Comment by evan-charmworks** [comment]
The CI tests the tests folder, but we should make sure examples and benchmarks pass too.

I noticed the privatization test says `No privatization methods supported for current AMPI target.` even though fsglobals should be supported on Windows.

**Comment by evan-charmworks** [comment]
Failed on my local Cygwin install. [cmake-win 20200630.txt](https://github.com/UIUC-PPL/charm/files/4860987/cmake-win.20200630.txt)

**Comment by matthiasdiener** [comment]
> cmake-win 20200630.txt

Relevant error:

```bash
$ ./build all-test multicore-win-x86_64 -j4 -g3
[...]
C:\msys64\home\Evan\charm\src\libs\ck-libs\armci\armci_impl.h(118): error C2084: function 'float fabsf(float)' already has a body
C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt\corecrt_math.h(670): note: see previous definition of 'fabsf'
C:\msys64\home\Evan\charm\src\libs\ck-libs\armci\armci_vp.C(21): error C2264: 'fabsf': error in function definition or declaration; function not called
Fatal Error by charmc in directory /cygdrive/c/msys64/home/Evan/charm/multicore-win-x86_64/src/libs/ck-libs
   Command /cygdrive/c/msys64/home/Evan/charm/multicore-win-x86_64/bin/unix2nt_cc -D_CRT_SECURE_NO_DEPRECATE -I/cygdrive/c/msys64/home/Evan/charm/multicore-win-x86_64/bin/../include -D__CHARMC__=1 -DAMPI_ERRHANDLER_RETURN=0 -g3 -g -U_FORTIFY_SOURCE -fno-stack-protector -c /cygdrive/c/msys64/home/Evan/charm/src/libs/ck-libs/armci/armci_vp.C -o CMakeFiles/modulearmci.dir/armci/armci_vp.C.o returned error code 2
charmc exiting...
```

**Comment by matthiasdiener** [comment]
> > cmake-win 20200630.txt
> 
> Relevant error:
> 
> ```shell
> $ ./build all-test multicore-win-x86_64 -j4 -g3
> [...]
> C:\msys64\home\Evan\charm\src\libs\ck-libs\armci\armci_impl.h(118): error C2084: function 'float fabsf(float)' already has a body
> C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\ucrt\corecrt_math.h(670): note: see previous definition of 'fabsf'
> C:\msys64\home\Evan\charm\src\libs\ck-libs\armci\armci_vp.C(21): error C2264: 'fabsf': error in function definition or declaration; function not called
> Fatal Error by charmc in directory /cygdrive/c/msys64/home/Evan/charm/multicore-win-x86_64/src/libs/ck-libs
>    Command /cygdrive/c/msys64/home/Evan/charm/multicore-win-x86_64/bin/unix2nt_cc -D_CRT_SECURE_NO_DEPRECATE -I/cygdrive/c/msys64/home/Evan/charm/multicore-win-x86_64/bin/../include -D__CHARMC__=1 -DAMPI_ERRHANDLER_RETURN=0 -g3 -g -U_FORTIFY_SOURCE -fno-stack-protector -c /cygdrive/c/msys64/home/Evan/charm/src/libs/ck-libs/armci/armci_vp.C -o CMakeFiles/modulearmci.dir/armci/armci_vp.C.o returned error code 2
> charmc exiting...
> ```

This error and the fsglobals error should be fixed now.

File: src/util/boost-context/CMakeLists.txt (hunk 5 above)
Comment (matthiasdiener) [reply src/util/boost-context/CMakeLists.txt:6]: I think all our ppc targets are ppc64le (ie, Power8 and later).

File: CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (matthiasdiener) [reply CMakeLists.txt]: Yeah, good point. I changed this to disable zlib only with MSVC.

**Comment by matthiasdiener** [comment]
> The CI tests the tests folder, but we should make sure examples and benchmarks pass too.
 Added the examples and benchmarks. Let's see if we don't hit timeouts.

**Comment by evan-charmworks** [comment]
I can confirm that all tests pass under a Cygwin environment. I tried getting MSYS2 to work and pushed various fixes for what I found, but it still doesn't work because `${CMAKE_COMMAND} -E create_symlink` fails on directories with MSYS2's CMake packages as well as Visual Studio's CMake.

If my additions look good, I will approve the PR.

File: cmake/detect-features.cmake [hunk 7]
```diff
@@ -65,15 +65,15 @@ set(CMAKE_REQUIRED_FLAGS "")
...
+elseif(CMK_HAS_DLOPEN AND CMAKE_SYSTEM_NAME STREQUAL "Darwin")
   set(CMK_SUPPORTS_FSGLOBALS 1)
 elseif(CMK_HAS_DLOPEN AND (${CMK_HAS_READLINK} OR ${CMK_HAS_REALPATH}))
   set(CMK_SUPPORTS_FSGLOBALS 1)
 else()
   set(CMK_SUPPORTS_FSGLOBALS 0)
 endif()
 
-if(${CMAKE_SYSTEM_NAME} STREQUAL "Windows")
+if(CMAKE_SYSTEM_NAME STREQUAL "Windows" OR CMAKE_SYSTEM_NAME STREQUAL "CYGWIN")
```
Comment (matthiasdiener) [inline cmake/detect-features.cmake:76]: I would put `if(CMK_WINDOWS)` here, like above.

**Comment by matthiasdiener** [comment]
Apart from the single comment, LGTM.

**Comment by evan-charmworks** [comment]
benchmarks/ampi/alltoall/mpibench.c need update to work on Windows. Failed to include sys/time.h.

**Comment by nitbhat** [comment]
> benchmarks/ampi/alltoall/mpibench.c needs update to work on Windows. Failed to include sys/time.h.

I replaced the old timer calls with `MPI_Wtime` in the latest patches.

**Pull request merged**